async boxers / unboxers

mixmix opened the pull request:
WIP:

:fire: This is a breaking change

## Motivation
I need to be able to do async lookups to look for appropriate keys for private-groups. 
This requires changing how boxers / unboxers work from sync > async.



---

## Questions

what in the ecosystem is currently using these APIs `box` / `unbox`, `addBoxer` / `addUnboxer`?

## TODO

if we're cutting a major release I intend to extract the box1 boxers/ unboxers so that they are optionally added as opposed to always on.

christianbundy commented on README.md:
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
 where
-- `unboxKey(ciphertext) => msgKey` is a function which tries to extract the message key from the encrypted content (`ciphertext`)
-- `unboxValue(ciphertext, msgKey) => plaintext` is a function which takes a message key and uses it to try to extract the message content from the `ciphertext`
+- `unboxKey(msg.value.content, msg.value, cb)`
```
Nit: Might be nice to be consistent here -- the `box()` signature references `content` and `recps` but this one references `msg.value.content` and `msg.value`.

mixmix replied:
true, if you think reading it like `msg.value.content` etc helps then I think making it consistent is a good idea

mixmix commented on minimal.js:
```diff
@@ -95,7 +95,15 @@ module.exports = function (dirname, keys, opts) {
+      if (err) {
+        console.error(err)
+        return cb(null, msg)
+      }
```
I'd like input thinking about how to handle error message with boxing / unboxing.

mixmix replied:
This here allows errors to bubble up through callbacks but then pops a log out and "returns" an unecrypted message.

christianbundy commented on README.md:
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
+- `unboxKey(msg.value.content, msg.value, cb)`
+  - is a function which tries to extract the message key from the encrypted content (`ciphertext`).
+  - is expected to callback `cb(err, msgKey)` which `msgKey` is the key for the message
+- `unboxValue(msg.value.content, msgKey, msg.value, cb)`
```
I'm sure there's a good reason why, but I don't understand:  could this just be `unboxValue(msg, cb)`? Seems like we've destructured `msg` into its parts and are passing `content` twice (in `msg.value.content` and `msg.value`).

mixmix replied:
I'm going to copy this comment into a comment I made about this.
TL;DR I agree

mixmix commented on minimal.js:
```diff
@@ -226,38 +239,68 @@ module.exports = function (dirname, keys, opts) {
+        break
+
+      default: throw new Error('invalid unboxer')
+    }
```
I decided to guard against weird stuff being poked in here

mixmix commented on minimal.js:
```diff
@@ -226,38 +239,68 @@ module.exports = function (dirname, keys, opts) {
+      },
+      value: function (ciphertext, msgKey, value, cb) {
+        cb(null, ssbKeys.unboxBody(ciphertext, msgKey))
+      }
```
I'd like a review on the signatures of these boxers / unboxers.
What adding a `cb` arg to the end surfaced is that the current signature has args which aren't being used (i.e. the `value = msg.value`).

I am tempted to refactor these functions to have signatures like : 

```js
function unboxValue (ciphertext, msgKey, context, cb) {
  // ...
}
```
where `context = { msg, ...futureThings }`

I'd also change the `unboxer` signature to `unboxer = { unboxKey: Function, unboxValue: Function }`

mixmix replied:
@christianbundy said : 

> I'm sure there's a good reason why, but I don't understand:  could this just be `unboxValue(msg, cb)`? Seems like we've destructured `msg` into its parts and are passing `content` twice (in `msg.value.content` and `msg.value`).

mixmix replied:
reviewed and pruned with Dominic

mixmix commented on create.js:
```diff
@@ -83,19 +83,18 @@ module.exports = function create (path, opts, keys) {
-          result = u.originalValue(data.value)
+            if (meta) cb(null, { key, value: result.value, timestamp: result.timestamp })
+            else cb(null, result.value)
+          })
```
I honestly had a hard time reading the original code because the logical paths through it seem woven and unclear to me.

I _think_ what I've written does the same thing. Certainly the tests present pass

christianbundy commented on create.js:
```diff
@@ -83,19 +83,18 @@ module.exports = function create (path, opts, keys) {
-          result = u.originalValue(data.value)
+            if (meta) cb(null, { key, value: result.value, timestamp: result.timestamp })
+            else cb(null, result.value)
+          })
```
Cool! Looks like we're only unboxing when we need to, which is great. Have you ran benchmarks on this code? It's important that it doesn't slow anything down, because this is a very hot path, but I have a hunch that this might even speed up methods like `createHistoryStream()`.

mixmix replied:
From call

use https://github.com/ssbc/bench-ssb

run 1-6
---

run these benches before + after on my system

mixmix replied:
## Process

- clone down [`bench-ssb`](https://github.com/ssbc/bench-ssb), npm linking `ssb-db`
- ran `run-all.sh` ... saw output turn up, but bench didn't get beyond test 9 for some reason (so I quit out)
- changed branch of `ssb-db` (back to master)
- changed `run-all.sh` to 
  - skip generate (so using same output.json as first test)
  - append `>>` results, instead of overwrite `>` (so can easily compare)
  - skipped tests `>= 9`

## Results

here's a sampling of results from tests 1-8
```
// 04-out.txt
minimal 100001 1 7197.941409342834 13.893 // async-box branch
minimal 100001 1 6958.527590285992 14.371 // master branch
```
```
07-out.txt
clock-dump 99001 9001 14375.054450413823 6.887  // async-box branch
clock-dump 99001 9001 14629.968967045957 6.767  // master branch
```

```
// 08-out.txt
sbot-read 100002 2 10718.327974276528 9.33  // async-box branch
sbot-read 100002 2 9706.105017955935 10.303 // master branch
```

christianbundy commented on minimal.js:
```diff
@@ -226,38 +239,68 @@ module.exports = function (dirname, keys, opts) {
-    if (recps.length === 0) throw new Error('private message field "recps" requires at least one recipient')
+    if (typeof recps === 'string') recps = content.recps = [recps]
+    if (!Array.isArray(recps)) return cb(new Error('private message field "recps" expects an Array of recipients'))
+    if (recps.length === 0) return cb(new Error('private message field "recps" requires at least one recipient'))
```
Do you think it would be wise to move the `if (!recps.every(isFeed)) return cb(null, null)` code here? Seems like that would be valid for any boxer, not just the default `box1` boxer, but I'm not confident about that.

mixmix replied:
No this is going to be different with different boxers. The new private groups for example is gonna let you do

```
  recps: [ @<base64 feedId>.sha256, %<base64 groupId>.cloaked ]
```

mixmix commented:
## Update

reviewed this with @dominictarr this afternoon. He's thought of some good changes - TL;DR the methods won't be async but we are extending the API.
Changes to follow

mixmix commented:
I think this is ready for a review if anyone wants to take a look.

christianbundy commented:
Is it pretty straightforward, or are there interesting changes that you made since we last discussed? I can probably take a look at this tomorrow sometime.

mixmix commented:
hmm, changes: 
- it's now sync again
  - dominic doesn't want to tempt people to do I/O on box / unbox
  - we figure a lot of keys state is small enough it can / should be in memory
  - we added an async `init` option the `addUnboxer`. This is where people can set state before things kick off
- refactoring
  - pulled out the auto box / unbox into `autobox.js`. way more chill IMO
  - wrote `AsyncJobQueue` to manage trickyness to knowing when things are ready or should wait. (with tests). collects logic nicely IMO
- removed `db.box` from the API exposed by `ssb-db`
  - this is now the only breaking change

mixmix commented on test/util.js:
```diff
@@ -1,12 +1,92 @@
-  opts.keys = opts.keys || ssbKeys.generate()
-  return require('../create')(dir, opts, opts.keys)
-}
+'use strict'
```
this file now tests `/util.js` and the original file has been moved to `test/create-ssb.js`

mixmix commented on test/box-unbox.js:
```diff
@@ -0,0 +1,286 @@
+'use strict'
```
this file used to be called `end-to-end.js`. I thought the name was misleading.
It's been renamed, and it's ~85% the same. I had to change some parts, and refactored a little as I went to be able to parse what the tests were doing

Importantly I've added tests around : 
- `addBoxer`
- `addUnboxer` (both signatures, and new async `init`)

arj03 commented on README.md:
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
 
-`box` is a function of signature `box(content, recps) => (String|null)`
-which is expected to either box the the message content and return a ciphertext String, or return null if it unable to.
+`box` is a function of signature `box(content, recps) => ciphertext` which is expected either:
```
Think this should be `boxer`?

arj03 commented on README.md:
```diff
@@ -450,26 +450,38 @@ an [observable](https://github.com/dominictarr/obv) of the current log sequence.
+  - is a function which tries to extract the message key from the encrypted content (`ciphertext`).
+  - is expected to return `msgKey` which is the key for the message
+- `unboxValue(msg.value.content, msgKey) => plainContent`
+  - is a function which takes a `msgKey` and uses it to try to extract the `plainContent` from the `ciphertext- `initBoxer(done)`
```
indention?

mixmix replied:
I have applied more teeth  

_indention !== indentation_ :heart:

arj03 commented on minimal.js:
```diff
@@ -10,81 +10,7 @@ var ssbKeys = require('ssb-keys')
-function isString (s) {
-  return typeof s === 'string'
-}
+var { box, unbox } = require('./autobox')
```
:heart:

arj03 commented on util.js:
```diff
@@ -159,3 +159,50 @@ exports.metaBackup = (msgValue, property) => {
   return msgValue.meta
 }
+
+exports.AsyncJobQueue = class AsyncJobQueue {
```
Someone must have made a module for this already?

mixmix replied:
yes, this started out as a minimal refactor and as I added tests got more complex.
Given it's working and working (I think) exactly like the code which was previously there, I'm kinda keen to leave this as is.

Feel free anyone to drop an alternative in and see if it passes the same tests!

arj03 commented:
Made a rather quick review. I havn't run any of this code. Was only reading the changes.

staltz commented:
I like to review files in a meaningful order, not the lexicographic order, do you think this order is meaningful (sorted from important to auxiliary):

```
 README.md               |  26 +++++--
 test/box-unbox.js       | 286 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 test/create-ssb.js      |  12 ++++
 test/close.js           |  22 +++---
 test/util.js            | 104 +++++++++++++++++++++++----
 test/end-to-end.js      | 170 --------------------------------------------
 autobox.js              |  84 ++++++++++++++++++++++
 minimal.js              | 257 +++++++++++++++++++++++++++----------------------------------------
 create.js               |  17 ++---
 index.js                |   1 -
 util.js                 |  47 +++++++++++++
 test/add.js             |   2 +-
 test/feed.js            |   2 +-
 test/history.js         |   2 +-
 test/latest.js          |   2 +-
 test/links.js           |   2 +-
 test/linktypes.js       |   2 +-
 test/log.js             |   2 +-
 test/mesages-by-type.js |   2 +-
 test/msg-encoding.js    |   2 +-
 test/validation.js      |   2 +-
 test/write-stream.js    |   2 +-
```

mixmix commented:
@staltz I'll split out a different PR which just moves and renames somethings which will reduce noise here

mixmix commented:
There's an intermitent fail in `test/close.js`. Not my fault I am 99% sure.

@staltz this should be easier to review.
The guts of the changes are in `minimal.js`

christianbundy commented on autobox.js:
```diff
@@ -0,0 +1,84 @@
+  var ciphertext
+  for (var i = 0; i < boxers.length; i++) {
+    const boxer = boxers[i]
+    ciphertext = boxer(content, recps)
```
Since we're doing a breaking change anyway, it'd be cool to drop this. Discussed with Mix. He told me to type this.

mixmix replied:
DONE

mixmix commented:
@christianbundy the changes we talked about have been added.
I'm merging this into master (but we'll delay publishing a major version till we get those close tests passing)
